runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/ElementCountMonitoringInfoToCounterUpdateTransformer.java

```diff
@@ -0,0 +1,119 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.fn.control;
+
+import com.google.api.services.dataflow.model.CounterUpdate;
+import com.google.api.services.dataflow.model.NameAndKind;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.beam.model.pipeline.v1.MetricsApi.MonitoringInfo;
+import org.apache.beam.runners.core.metrics.SpecMonitoringInfoValidator;
+import org.apache.beam.runners.dataflow.worker.counters.DataflowCounterUpdateExtractor;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * MonitoringInfo to CounterUpdate transformer capable to transform ElementCount counters.
+ */
+public class ElementCountMonitoringInfoToCounterUpdateTransformer
+    implements MonitoringInfoToCounterUpdateTransformer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BeamFnMapTaskExecutor.class);
+
+  private final SpecMonitoringInfoValidator specValidator;
+  private final Map<String, String> sdkToDfePCollectionMapping;
+  private static final String supportedUrn = "beam:metric:element_count:v1";
+
+  /**
+   * @param specValidator SpecMonitoringInfoValidator to utilize for default validation.
+   * @param sdkToDfePCollectionMapping Mapping of SDK available PCollection name to DFE pcollection
+   *     name.
+   */
+  public ElementCountMonitoringInfoToCounterUpdateTransformer(
+      SpecMonitoringInfoValidator specValidator, Map<String, String> sdkToDfePCollectionMapping) {
+    this.specValidator = specValidator;
+    this.sdkToDfePCollectionMapping = sdkToDfePCollectionMapping;
+  }
+
+  /**
+   * Validates provided monitoring info against specs and common safety checks.
+   *
+   * @param monitoringInfo to validate.
+   * @return Optional.empty() all validation checks are passed. Optional with error text otherwise.
+   * @throws RuntimeException if received unexpected urn.
+   */
+  protected Optional<String> validate(MonitoringInfo monitoringInfo) {
+    Optional<String> validatorResult = specValidator.validate(monitoringInfo);
+    if (validatorResult.isPresent()) {
+      return validatorResult;
+    }
+
+    String urn = monitoringInfo.getUrn();
+    if (!urn.equals(supportedUrn)) {
+      throw new RuntimeException(String.format("Received unexpected counter urn: %s", urn));
+    }
+
+    // todomigryz: extract and utilize pcollection label from beam_fn_api.proto
+    if (!sdkToDfePCollectionMapping.containsKey(monitoringInfo.getLabelsMap().get("PCOLLECTION"))) {
+      return Optional.of(
+          "Encountered ElementCount MonitoringInfo with unknown PCollectionId: "
+              + monitoringInfo.toString());
+    }
+
+    return Optional.empty();
+  }
+
+  @Override
+  public CounterUpdate transform(MonitoringInfo monitoringInfo) {
```

Comment:
Please document the steps required to transform this properly. This is a bit convoluted, so it would help to
describe what your starting with, what your outputting and what the helper maps contain.

Please include a comment with some example strings based on the dumps you showed me.